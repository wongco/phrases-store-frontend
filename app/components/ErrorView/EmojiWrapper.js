import styled from 'styled-components';

const EmojiWrapper = styled.div`
  font-size: 150%;
  margin: 0 0.5em;
`;

export default EmojiWrapper;
