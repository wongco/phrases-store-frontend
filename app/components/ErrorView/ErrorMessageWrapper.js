import styled from 'styled-components';

const ErrorMessageWrapper = styled.div`
  margin: 0.5em;
`;

export default ErrorMessageWrapper;
