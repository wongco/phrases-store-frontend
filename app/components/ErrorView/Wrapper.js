import styled from 'styled-components';

const Wrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  margin: 1em 0;
`;

export default Wrapper;
