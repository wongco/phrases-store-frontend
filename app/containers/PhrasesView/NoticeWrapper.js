import styled from 'styled-components';

const NoticeWrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
`;

export default NoticeWrapper;
