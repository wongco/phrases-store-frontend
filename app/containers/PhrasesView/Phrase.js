import React from 'react';
import PropTypes from 'prop-types';

const Phrase = props => {
  const { text } = props;
  return <li>{text}</li>;
};

Phrase.propTypes = {
  text: PropTypes.string,
};

export default Phrase;
