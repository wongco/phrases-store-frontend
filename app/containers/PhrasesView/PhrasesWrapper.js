import styled from 'styled-components';

const PhraseWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  width: 80%;
`;

export default PhraseWrapper;
