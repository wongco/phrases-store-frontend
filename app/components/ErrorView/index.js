import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Wrapper from './Wrapper';
import EmojiWrapper from './EmojiWrapper';
import ErrorMessageWrapper from './ErrorMessageWrapper';

class ErrorView extends PureComponent {
  static Image = props => <EmojiWrapper>{props.emoji}</EmojiWrapper>;

  static Message = props => (
    <ErrorMessageWrapper>{props.children}</ErrorMessageWrapper>
  );

  render() {
    return <Wrapper>{this.props.children}</Wrapper>;
  }
}

ErrorView.propTypes = {
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
};

export default ErrorView;
