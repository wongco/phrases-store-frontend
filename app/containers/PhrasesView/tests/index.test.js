import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow, mount } from 'enzyme';
import { enzymeFind } from 'styled-components/test-utils';
import { IntlProvider } from 'react-intl';

import ErrorView from 'components/ErrorView';
import LoadingIndicator from 'components/LoadingIndicator';
import { PhrasesView } from '../index';
import PhrasesWrapper from '../PhrasesWrapper';
import Phrase from '../Phrase';
import NoticeWrapper from '../NoticeWrapper';

describe('<PhrasesView />', () => {
  const phrases = [
    {
      id: 1,
      text: 'first phrase!',
    },
    {
      id: 2,
      text: 'second phrase!',
    },
  ];
  const mockLoadPhrases = jest.fn();

  it('renders without crashing', () => {
    shallow(<PhrasesView phrases={phrases} loadPhrases={mockLoadPhrases} />);
  });

  it('matches snapshot for default criteria', () => {
    const wrapper = shallow(
      <PhrasesView phrases={phrases} loadPhrases={mockLoadPhrases} />,
    );
    const serialized = toJson(wrapper);
    expect(serialized).toMatchSnapshot();
  });

  describe('rendering tests', () => {
    it('renders imported components normally on load', () => {
      const wrapper = shallow(
        <PhrasesView phrases={phrases} loadPhrases={mockLoadPhrases} />,
      );
      const PhrasesWrapperComponent = wrapper.find(PhrasesWrapper);
      expect(PhrasesWrapperComponent).toHaveLength(1);
      const PhraseComponent = wrapper.find(Phrase);
      expect(PhraseComponent).toHaveLength(2);
    });

    it('should not render ErrorView or LoadingIndicator initially', () => {
      const wrapper = shallow(
        <PhrasesView
          phrases={phrases}
          loadPhrases={mockLoadPhrases}
          loading={false}
          error={false}
        />,
      );
      const loadingComponent = wrapper.find(LoadingIndicator);
      expect(loadingComponent).toHaveLength(0);
      const errorView = wrapper.find(ErrorView);
      expect(errorView).toHaveLength(0);
    });

    it('when loading is true, renders LoadingIndicator, no phrases rendered', () => {
      const wrapper = shallow(
        <PhrasesView phrases={phrases} loadPhrases={mockLoadPhrases} loading />,
      );
      const loadingComponent = wrapper.find(LoadingIndicator);
      expect(loadingComponent).toHaveLength(1);
      const PhraseComponent = wrapper.find(Phrase);
      expect(PhraseComponent).toHaveLength(0);
    });

    it('when error is true, renders ErrorView', () => {
      const wrapper = shallow(
        <PhrasesView phrases={phrases} loadPhrases={mockLoadPhrases} error />,
      );
      const errorView = wrapper.find(ErrorView);
      expect(errorView).toHaveLength(1);
    });

    it('render notice if there are no phrases on api', () => {
      const mockPhrasesEmpty = [];
      const wrapper = mount(
        <IntlProvider locale="en">
          <PhrasesView
            phrases={mockPhrasesEmpty}
            loadPhrases={mockLoadPhrases}
          />
        </IntlProvider>,
      );
      const NoticeWrapperComponent = enzymeFind(wrapper, NoticeWrapper);
      expect(NoticeWrapperComponent).toHaveLength(1);
    });
  });

  describe('lifecycle method tests', () => {
    it('upon mounting, loadPhrases is called', () => {
      const loadPhrasesLocal = jest.fn();
      shallow(<PhrasesView phrases={phrases} loadPhrases={loadPhrasesLocal} />);

      expect(loadPhrasesLocal).toHaveBeenCalled();
    });

    it('upon unmount, clearAppError is called', () => {
      const mockClearAppError = jest.fn();
      const wrapper = shallow(
        <PhrasesView
          phrases={phrases}
          loadPhrases={mockLoadPhrases}
          clearAppError={mockClearAppError}
        />,
      );
      wrapper.unmount();

      expect(mockClearAppError).toHaveBeenCalled();
    });
  });
});
