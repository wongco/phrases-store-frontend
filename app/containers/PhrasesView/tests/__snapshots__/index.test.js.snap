// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhrasesView /> matches snapshot for default criteria 1`] = `
<PhraseViewWrapper>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Show All Phrases
    </title>
    <meta
      content="Description of PhrasesView"
      name="description"
    />
  </HelmetWrapper>
  <H1>
    <FormattedMessage
      defaultMessage="Ordered by most recent:"
      id="app.containers.PhrasesView.header"
      values={Object {}}
    />
  </H1>
  <PhrasesWrapper__PhraseWrapper>
    <ul>
      <Phrase
        key="1"
        text="first phrase!"
      />
      <Phrase
        key="2"
        text="second phrase!"
      />
    </ul>
  </PhrasesWrapper__PhraseWrapper>
</PhraseViewWrapper>
`;
